payment confirmation view standardized for payment gateway

ravinderk opened the pull request:
## Description
Resolves: #4637 
<!-- Please describe your changes -->

This pr will add a way to confirm payment for embed forms.

## Affects
<!-- Provide a list of areas in the code base affected by this. Not file by file, just descriptively. -->
It will affect the receipt loading process because if the developer sets donation to confirmation when the donor returns back to the website then before showing receipt we can show the donation confirmation view. Donation confirmation view will only load when a donation is pending but it always is hidden behind load for Sequoia form template.

I add an action hook to confirm donation before loading the donation confirmation view `give_handle_donation_confirmation`.

## What to test
<!-- Provide a comprehensive list of what should be tested to confirm this works and not break anything. -->
You can process the payment with the `PayPal standard` gateway. After payment confirmation when return to the website then you will wait for `5 sec` to view receipt if the donation does not complete yet ( that means we did not get `ipn` from PayPal)

## Screenshots:
<!-- Optional. Any screenshots or animated gifs that may help others understand the changes -->
![image](https://user-images.githubusercontent.com/1784821/80012263-a85b7080-84ea-11ea-9e5a-94e43cacb924.png)


## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

ravinderk commented:
@JasonTheAdams You can ignore newly created helper functions. I will create another issue to refactor helper function.

DevinWalker commented on src/Controller/Form.php:
```diff
@@ -87,10 +90,30 @@ public function loadReceiptView() {
+					/**
+					 * Fire the action hook.
+					 *
+					 * If developer want to verify payment before showing receipt then use `give_handle_donation_confirm` action hook to verify donation.
```
Please clean up grammar: `If a developer wants`...

ravinderk replied:
https://github.com/impress-org/givewp/compare/a8cb87580b81...c652b8612198

DevinWalker commented on src/Controller/Form.php:
```diff
@@ -87,10 +90,30 @@ public function loadReceiptView() {
+
+					$donation = new Donation( $donationId );
+
+					// Load payment processing video only if donation is in pending status.
```
Video?

ravinderk replied:
it is `view`.

ravinderk replied:
https://github.com/impress-org/givewp/compare/c652b8612198...1a732f01cf52

DevinWalker commented on src/Helpers/Session/Donation.php:
```diff
@@ -0,0 +1,17 @@
+use function give_get_donation_id_by_key as getDonationIdByPurchaseKey;
+
+/**
+ * Get donation is from donor session.
```
ID, not "is" I'm assuming

ravinderk replied:
https://github.com/impress-org/givewp/compare/1a732f01cf52...04064e2ec6fd

JasonTheAdams commented on composer.json:
```diff
@@ -40,7 +40,10 @@
             "src/Helpers/Form/Utils.php",
             "src/Helpers/Frontend/Shortcode.php",
-            "src/Helpers/Utils.php"
+            "src/Helpers/Utils.php",
+            "src/Helpers/Session/Session.php",
+            "src/Helpers/Session/Donation.php",
+            "src/Helpers/Session/DonationConfirmation/Frontend.php"
```
We should really start to build out static facades sooner than later. We're loading _all_ of these files on every request and the `use function ...` at the top of the files is getting crazy.

ravinderk replied:
@JasonTheAdams Helper function which is added to this pr converted into there own class and facades. I will create separate issue to convert other helper functions.

JasonTheAdams commented on src/Donation/Donation.php:
```diff
@@ -0,0 +1,31 @@
+ *
+ * @package Give\Donation
+ */
+class Donation {
```
What is your intention for this class? Is this a model?

ravinderk replied:
@JasonTheAdams Use this is a modal. I added basic as per requirement. Let me know if you want suggest to add more properties.

ravinderk replied:
@JasonTheAdams Donation model requires discussion and research because currently, it stores in WP `post` table. For now, I remove it: https://github.com/impress-org/givewp/commit/0c4d39db3ea31828d757fd0eb64b503d85950cb2

JasonTheAdams commented on src/Helpers/Session/DonationConfirmation/Frontend.php:
```diff
@@ -0,0 +1,58 @@
+	$paymentGatewayId = ucfirst( give_clean( $_GET['payment-confirmation'] ) );
+	removeDataFromSession( "postDataFor{$paymentGatewayId}" );
+
+	return false;
```
Why are you returning false? If you don't need to return then just let it implicitly return void.

ravinderk replied:
I do not make sense to`return` any, so removing this.

ravinderk replied:
https://github.com/impress-org/givewp/compare/04064e2ec6fd...ddd8d96e937e

JasonTheAdams commented on src/Helpers/Session/DonationConfirmation/Frontend.php:
```diff
@@ -0,0 +1,58 @@
+use function Give\Helpers\Form\Utils\isConfirmingDonation;
+
+/**
+ * Store posted data to donor session to access it in iframe if we are on payment confirmation page.
```
Please explain when it returns true/false.

JasonTheAdams commented on src/Helpers/Session/Session.php:
```diff
@@ -0,0 +1,58 @@
+function storeDataIntoSession( $key, $data, $replace = true ) {
+	$session = getSession();
+
+	if ( null === $data ) {
```
Careful that this isn't creating an intended effect, here. I notice that `removeDataFromSession` just calls this and passes null. Is there a scenario where you would actually want to store null in the session? The null value here has an implicit result. Maybe that's fine, you just have to be sure then scenario you're using (i.e. data is null) will never happen and _not_ intend to remove the key. Make sense?

ravinderk replied:
@JasonTheAdams nothing will be store in session but instead if pass `null` as data value it will remove that key from session `array`. In `if` condition, I am using `unset` to remove data.

JasonTheAdams commented on src/Views/Form/Templates/Sequoia/views/receipt.php:
```diff
@@ -4,22 +4,23 @@
-$payment = new Payment( getPaymentId() );
-$options = getTemplateOptions();
+/* @var Payment $donation */
+$donation = new Payment( getPaymentId() );
```
I notice that you just replaced `$payment` with `$donation`. What's the reason? Your dealing with an instance of `Payment`, so isn't `$payment` a more appropriate name?

ravinderk replied:
@JasonTheAdams In Give, we do not use payment instead we use `donation`. payment term still exists because code was forked.

ravinderk commented:
@JasonTheAdams I refactored pr. I created two new classed `Give\Session\Access` (`abstract`) and `Give\Session\Access\Donation`.

You can see the data that we store in the donation session. I did not create object properties to access these properties, you can access them from an array. I thought that it is not part of scope and take little more time to develop that why I did not add logic for this.

Data in donation session:
```
Array
(
    [price] => 3000
    [purchase_key] => b2a75d6a0b0ca408c76d15cc3558c143
    [user_email] => mehul@test.com
    [date] => 2020-04-23 00:02:41
    [user_info] => Array
        (
            [id] => 12
            [title] => 
            [email] => mehul@test.com
            [first_name] => Mehul
            [last_name] => Gohil
            [address] => 
        )

    [post_data] => Array
        (
            [give-fee-amount] => 89.91
            [give-fee-mode-enable] => false
            [give-fee-status] => enabled
            [give-honeypot] => 
            [give-form-id-prefix] => 1398-1
            [give-form-id] => 1398
            [give-form-title] => MailChimp 69
            [give-current-url] => http://give.test/donations/mailchimp-69/
            [give-form-url] => http://give.test/givewp/mailchimp-69/
            [give-form-minimum] => 100
            [give-form-maximum] => 99999999
            [give-form-hash] => db56ec5851
            [give-price-id] => 2
            [give-amount] => 3.000
            [give-recurring-period] => on
            [give-recurring-period-donors-choice] => quarter
            [give_stripe_payment_method] => 
            [give-recurring-logged-in-only] => 
            [give-logged-in-only] => 1
            [_give_is_donation_recurring] => 1
            [give_recurring_donation_details] => {"give_recurring_option":"yes_donor"}
            [give-fee-recovery-settings] => {"fee_data":{"all_gateways":{"percentage":"2.900000","base_amount":"0.300000","give_fee_disable":false,"give_fee_status":true,"is_break_down":true}},"give_fee_status":true,"give_fee_disable":false,"is_break_down":true,"fee_mode":"donor_opt_in","is_fee_mode":true,"fee_recovery":true}
            [give_first] => Mehul
            [give_last] => Gohil
            [give_company_name] => impress.org
            [give_email] => mehul@test.com
            [payment-mode] => stripe_checkout
            [give_action] => purchase
            [give-gateway] => stripe_checkout
            [give_razorpay_form_merchant_id] => rzp_test_wsz50mgHQecPuE
            [give_embed_form] => 1
            [give-purchase] => Donate Now
        )

    [gateway] => stripe_checkout
    [card_info] => Array
        (
            [card_name] => 
            [card_cvc] => 
            [card_exp_month] => 
            [card_exp_year] => 
            [card_address] => 
            [card_address_2] => 
            [card_city] => 
            [card_state] => 
            [card_country] => 
            [card_zip] => 
        )

    [period] => quarter
    [times] => 0
    [frequency] => 1
    [donation_id] => 1682
)
```

Helper function created for this pr also converted to facades. 

Let me know if you have any questions.

ravinderk commented:
@JasonTheAdams I updated PR. Now `Give\Session\Access\Donation` return `object` type data.
Because in `object` I renamed properties, so it is complex to transform it back to `array` that why I am managing two properties, one will be used when store data and other will be used to return queried data ( `$data` & `$dataObj` in `Give\Session\Access` ).

We object will look like this:

```
Give\ValueObjects\Session\Donation Object
(
    [id] => 1729
    [totalAmount] => 20
    [purchaseKey] => 6d964fe6339c57106d889fc18f8d9ab1
    [donorEmail] => ravinder.kumar.apj@gmail.com
    [createdAt] => DateTime Object
        (
            [date] => 2020-04-30 22:51:37.000000
            [timezone_type] => 3
            [timezone] => UTC
        )

    [paymentGateway] => stripe
    [caseTo:Give\ValueObjects\Session\Donation:private] => Array
        (
            [formEntries] => Give\ValueObjects\Session\FormEntries
            [donorInfo] => Give\ValueObjects\DonorInfo
            [cardInfo] => Give\ValueObjects\CardInfo
        )

    [donorInfo] => Give\ValueObjects\DonorInfo Object
        (
            [wpUserId] => 1
            [email] => ravinder.kumar.apj@gmail.com
            [address] => 
            [firstName] => Ravinder
            [lastName] => 
            [honorific] => 
        )

    [formEntries] => Give\ValueObjects\Session\FormEntries Object
        (
            [formId] => 1369
            [formTitle] => Form Template
            [currentUrl] => http://give.test/donations/form-template/
            [priceId] => 1
            [totalAmount] => 20
            [firstName] => Ravinder
            [lastName] => Kumar
            [companyName] => GiveWP
            [donorEmail] => ravinder.kumar.apj@gmail.com
            [wpUserId] => 1
            [paymentGateway] => stripe
            [feeAmount] => 0.91
            [feeModeEnable] => false
            [feeStatus] => enabled
            [honeypot] => 
            [formIdPrefix] => 1369-1
            [formUrl] => http://give.test/givewp/form-template/
            [formMinimum] => 1
            [formMaximum] => 1000000
            [formHash] => 1333fa00cc
            [stripePaymentMethod] => pm_0Gdg4p8zP34B9bTNoFxObE7o
            [recurringLoggedInOnly] => 
            [loggedInOnly] => 1
            [recurringDonationDetails] => {"is_recurring":false}
            [feeRecoverySettings] => {"fee_data":{"all_gateways":{"percentage":"2.900000","base_amount":"0.300000","give_fee_disable":false,"give_fee_status":true,"is_break_down":true}},"give_fee_status":true,"give_fee_disable":false,"is_break_down":true,"fee_mode":"donor_opt_in","is_fee_mode":true,"fee_recovery":true}
            [paymentMode] => stripe
            [cardName] => Mehul Gohil
            [action] => purchase
            [razorpayFormMerchantId] => rzp_test_wsz50mgHQecPuE
            [embedForm] => 1
        )

    [cardInfo] => Give\ValueObjects\CardInfo Object
        (
            [name] => Mehul Gohil
            [number] => 
            [cvc] => 
            [expMonth] => 
            [expYear] => 
            [address] => Give\ValueObjects\Address Object
                (
                    [line1] => 
                    [line2] => 
                    [city] => 
                    [state] => 
                    [postalCode] => 
                    [country] => 
                )
        )
)

```

In the above object, you can see that `cardInfo`, `donorInfo`, and `formEntries` also has there own object because they will be present every time when the donation will create. Ohter unknown values will be converted to `stdClass` (non-recursively) if `array` type.

JasonTheAdams commented on src/Session/Access/Donation.php:
```diff
@@ -0,0 +1,87 @@
+ *
+ * @package Give\Session\Access
+ *
+ * @property-read string         $sessionKey
+ * @property-read DonationObject $dataObj
+ * @property-read array          $renameTo
```
You don't need to add these if the properties are fixed. These docblocks are only necessary for dynamic properties and methods.

ravinderk replied:
fixed: https://github.com/impress-org/givewp/compare/1e4adddcb256...2a7b11961de2

JasonTheAdams commented on src/ValueObjects/Session/FormEntries.php:
```diff
@@ -0,0 +1,121 @@
+use Give\ValueObjects\ValueObjects;
+use InvalidArgumentException;
+
+class FormEntries implements ValueObjects {
```
I like what you're doing here, but these aren't actually Value Objects. a Value Object is a pattern for when an object represents a _single_ value. For example, you may have a `Currency` class which represents a single integer value. This is really just a run-of-the-mill class. Your `ValueObjects` class can be renamed to something like `ArraySerializable` and have `fromArray` and `toArray` methods.

ravinderk replied:
@JasonTheAdams Agree with you. `ValueObjects`'s structure must be defined and immutable that means we can not overwrite them.  I find out that there is not any scope of these objects outside or session that why I am moving `FormEntry` and `Donation`to `Session\Objects`.

ref: https://github.com/impress-org/givewp/compare/2a7b11961de2...4bdd65fe7f79

ravinderk replied:
resolved

ravinderk commented:
@JasonTheAdams Directories has been renamed as discussed.

![image](https://user-images.githubusercontent.com/1784821/81000166-80adc600-8e63-11ea-938c-10e1a6e9600e.png)

DevinWalker commented:
@ravinderk can you please update the merge conflict for this branch.

JasonTheAdams commented on src/Session/SessionDonation/SessionObjects/Objects.php:
```diff
@@ -0,0 +1,12 @@
+<?php
+namespace Give\Session\SessionDonation\SessionObjects;
+
+interface Objects {
```
@ravinderk Let's move this to the `Give\Session` root and call it `SessionObject`. I suspect it will be useful in other domains.

ravinderk replied:
fixed: https://github.com/impress-org/givewp/compare/89e8238e2fa8...c14c8ac2fdf6

JasonTheAdams commented on src/Session/SessionDonation/SessionObjects/Donation.php:
```diff
@@ -0,0 +1,177 @@
+use Give\ValueObjects\ValueObjects;
+use InvalidArgumentException;
+
+class Donation implements Objects {
```
Please provide a docblock for this class which includes a comment explaining that this is a donation only as far as what it is being constructed in the session. It does not reflect a donation in the database. It's a pending donation.

ravinderk replied:
fixed: https://github.com/impress-org/givewp/compare/c14c8ac2fdf6...3dcd52854e4f
